test/App/Fossa/API/BuildWaitSpec.hs

```diff
@@ -0,0 +1,181 @@
+module App.Fossa.API.BuildWaitSpec (spec) where
+
+import App.Fossa.API.BuildWait (WaitConfig (WaitConfig, apiPollDelay), waitForBuild, waitForIssues, waitForScanCompletion)
+import Control.Algebra (Has)
+import Control.Carrier.Reader (ReaderC, runReader)
+import Control.Effect.FossaApiClient (FossaApiClientF (..))
+import Control.Effect.Lift (Lift)
+import Control.Monad (void)
+import Control.Timeout (Cancel, Duration (MilliSeconds, Seconds), timeout')
+import Data.Text (Text)
+import Fossa.API.Types (
+  Build (..),
+  BuildStatus (StatusCreated, StatusFailed, StatusRunning, StatusSucceeded),
+  BuildTask (..),
+  Issues (issuesStatus),
+  Project (projectIsMonorepo),
+  ScanResponse (responseScanStatus),
+ )
+import Srclib.Types (Locator (..))
+import Test.Effect (expectFatal', it', shouldBe')
+import Test.Fixtures qualified as Fixtures
+import Test.Hspec (Spec, describe)
+import Test.MockApi (MockApi, alwaysReturns, returnsOnce)
+
+runWithConfigAndTimeout ::
+  Has (Lift IO) sig m =>
+  (Cancel -> ReaderC WaitConfig m a) ->
+  m ()
+runWithConfigAndTimeout =
+  void
+    . runReader waitConfig
+    . timeout' (Seconds 1)
+
+spec :: Spec
```

Comment:
Can we put the helper functions _above_ the spec?  I find it's easier to understand tests when they're written in this order:
1. Simple-data-building functions (e.g.: `mkFixture a b = Foo (Bar b) . Just $ Baz a`)
2. Simple data (expectations and inputs)
3. Collections of simple data
4. Complex expectations and inputs (rare in haskell, but it does happen)
5. Trivial test helpers
6. Non-trivial test helpers
7. The actual tests

Reply:
Hmmm, I felt like maybe they weren't the interesting point and the names were descriptive enough.  I sometimes hate having to scroll through lots of data set up just to find the tests themselves.  It's sort of like having `where` blocks on a function.  Easy enough to flip, but it feels like burying the lede.

Reply:
Interesting.  I view the most important part of tests to be the description of behavior and the high-level steps taken to test it.  A user can then dig into the data if they want more information, but they probably don't care too much about the details of `expectGetApiOpts`.  Maybe I'm thinking that by naming these expectations they are clear enough that the details of setting the expectation doesn't have to be spelled out and the noise of selecting fixtures can be pushed down.

Reply:
Huh, I come at it from a totally different perspective.  I'm more behavior-driven-design oriented.

I first want to check that the right things are being tested, then I have the context to look at the data and make sure it's the right data.  Without context the data doesn't mean much to me.  Let's say one of the data set-up functions has an empty list.  Is that correct or not?  I don't feel like I have the context to evaluate the data until I've read the tests.  The tests show me intention, the data doesn't.

Is this just the case that I read tests in the opposite order you do?

This may also come from spending most of my time on tests reviewing them.  When I review tests the most important thing to me is that the author has good coverage of tests over the behavior of the code.  If something is missing or the tests are messy then I send it back without looking too much at the details.  Once we have the right tests then I can start to look at the details to see how they are setting things up and make sure it matches what they say they are doing.

Also, sometimes to understand what code does I'll go browse the test cases because they should describe the behavior.

Reply:
Frankly, these are both valid ways to read tests, so I'm not going to keep blocking you on this.  I also don't think this is something to standardize on, since it's clearly a very individual-perspective topic.

Reply:
It's interesting that we read these in polar opposite ways though.  I never would have expected that.